### PR TITLE
Add a doc explaining recompilation.

### PR DESCRIPTION
**Pull request by ailzhang**

### CODE REVIEW

**Review by JackCaoG — APPROVED** [review]
Great summary, thanks @ailzhang !

**Comment by JackCaoG** [comment]
@yaochengji @byronyi for visibility.

**Comment by ailzhang** [comment]
I'll merge this for now, happy to take more questions/comments in this thread and I can address them in followup PRs. :D

**Comment by byronyi** [comment]
> @yaochengji @byronyi for visibility.

Thanks. This is indeed a great doc to explain current pain points from users' point of view.

@ailzhang For 2.2 and 3, would it be a better UX if we have a `torch.jit.script` wrapper on the submodule that does those "very PyTorch-y" things? Currently XLA supports [if](https://www.tensorflow.org/xla/operation_semantics#conditional) and [while](https://www.tensorflow.org/xla/operation_semantics#while), and assuming XLA is going to support dynamic shape in its IR (e.g. MHLO), we still need PyTorch passing those information from frontend. Then the whole program could still use the lazy tensor approach (main training loop) while we might even solve the TS->XLA compilation AOT.

File: docs/source/notes/source_of_recompilation.md [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+* parse python source. code to get the control flow statement automatically. This is like Torchscript and somehow merge the torchscripted graph into the lazily trace graph properly (including shape info etc). I haven’t thought through the steps of how to implement this indeed :P
+
+But either solution above requires non-trivial amount of effort, either on user side or on the framework side. That’s why we currently just take the hit of early evaluation & multiple compilations as a short term solution given the bandwidth we have.
+
+Okay so now we assume that also have control flow lowered in the graph automagically, are we gold?
+
+YES! Now you have your whole computation represented in a graph of Tensor operations, including control flow so that compilers can now consume and do their smart tricks! But tbh at this point your program is no longer very PyTorch-y.
```
Comment (stellaraccident) [inline docs/source/notes/source_of_recompilation.md:115]: Ftr - TF itself takes this approach and there are a lot of expressivity gaps, even before you further reduce to HLO, which had even tighter constraints. It falls into the category of things that people do when they have no choice, but it is painful to watch users try to navigate the land mines :/

File: docs/source/notes/source_of_recompilation.md [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+XLA solves this problem in the other way, by using padding/bucketization (for #1) and bounded dynamic shape (for #2.1).
+
+Brian Hirsh(@bdhirsh) asked some really good questions in the comment, moving here to make them more visible:
+
+2. Is it worth sticking a TORCH_WARN in the XLA kernels of ops that produce data-dependent output shapes?
+
+Yea torch_warn is useful in telling users "hey your program won't run blazing fast". But for these data dependent ops, there isn't an easy rewrite for them unless users change the logic in their model. (another example is torch.unique())
+
+3. How ops like nonzero impact our ability to devirtualize sizes()? If we want to devirtualize sizes(), we’ll need to be able to eagerly compute sizes for each op - won’t that mean we’re forced to evaluate the graph every time we hit an op like nonzero? Vs. right now, it sounds like we don't actually force an evaluation when a user calls nonzero()?
```
Comment (byronyi) [inline docs/source/notes/source_of_recompilation.md:152]: @ailzhang Do you have a little bit more context for the intent to devirtualize sizes()? 

Leave torch/xla aside, current cuda implementation of nonzero actually triggers a host-device sync, and that cost would be more substantial than a devirtualization of a virtual function call.

**Pull request merged**